tensorflow_probability/python/experimental/mcmc/particle_filter.py

```diff
@@ -514,30 +660,42 @@ def _compute_observation_log_weights(step,
         ps.equal(step % num_transitions_per_observation, 0))
     observation_idx = step // num_transitions_per_observation
     observation = tf.nest.map_structure(
-        lambda x, step=step: tf.gather(x, observation_idx), observations)
+        lambda x, step=step: tf.gather(x,
+                                       observation_idx,
+                                       axis=particles_dim),
+        observations)
+
+    # Expand dimensions of observation at the particles_dim
+    observation = tf.expand_dims(observation, axis=particles_dim)
```

Comment:
NOTE: The goal of this expand_dims is something like:

 - `observations` has shape like `[num_steps] + batch_shape`.

 - so `observation` has shape like `batch_shape`.

 - particles has shape like `batch_shape_1 + [num_particles] + batch_shape_2 + event_shape`

 - `observation_fn(step, particles)` is a distribution with batch shape like `batch_shape_1 + [num_particles] + batch_shape_2` and event shape `event_shape`.

 - When particles_dim=0, we can just apply `observation_fn(step, particles).log_prob(observations)`, because the `log_prob` is required to broadcast the observations of shape `batch_shape + event_shape` up to `[num_particles] + batch_shape + event_shape`.

 - But when particles_dim>0, this won't work.  Instead, we have to expand_dims on `observations` so it has shape like `batch_shape_1 + [1] + batch_shape_2 + [event_shape`.  This will allow the `log_prob` to broadcast `observations` up to shape `batch_shape_1 + [num_steps] + batch_shape_2 + [event_shape]`, so that the `log_prob` can run and will return something with shape `batch_shape_1 + [num_steps] + batch_shape_2`.